mlir/lib/Dialect/Arith/Transforms/EliminateExplicitRounding.cpp

```diff
@@ -0,0 +1,86 @@
+//-===//
+//
+// Part of the LLVM Project, under the Apache License v2.0 with LLVM Exceptions.
+// See https://llvm.org/LICENSE.txt for license information.
+// SPDX-License-Identifier: Apache-2.0 WITH LLVM-exception
+//
+//===----------------------------------------------------------------------===//
+//
+// This file implements removing intermediate extf/truncf pairs inserted from
+// type conversion.
+//
+//===----------------------------------------------------------------------===//
+#include "mlir/Dialect/Arith/Transforms/Passes.h"
+
+#include "mlir/Dialect/Arith/IR/Arith.h"
+#include "mlir/IR/PatternMatch.h"
+#include "mlir/IR/TypeUtilities.h"
+#include "mlir/Transforms/GreedyPatternRewriteDriver.h"
+
+namespace mlir {
+namespace arith {
+#define GEN_PASS_DEF_ELIMINATEEXPLICITROUNDING
+#include "mlir/Dialect/Arith/Transforms/Passes.h.inc"
+} // namespace arith
+} // namespace mlir
+
+using namespace mlir;
+
+namespace {
+
+struct EliminateExplicitRoundingRewritePattern final
+    : OpRewritePattern<arith::ExtFOp> {
+  using OpRewritePattern::OpRewritePattern;
+  using FilterFunction = std::function<bool(Operation *)>;
+
+  EliminateExplicitRoundingRewritePattern(MLIRContext *context,
+                                          FilterFunction filterFunc = nullptr)
+      : OpRewritePattern(context), filterFunc(filterFunc) {}
+
+  LogicalResult matchAndRewrite(arith::ExtFOp extFOp,
+                                PatternRewriter &rewriter) const final {
+    // Check whether match `truncF->extF` pair.
+    auto truncFOp = extFOp.getOperand().getDefiningOp<arith::TruncFOp>();
+    if (!truncFOp)
+      return failure();
+
+    // Check whether need to filter out.
+    if (filterFunc && filterFunc(extFOp))
+      return failure();
+
+    // Check whether the rounding pair's input and output data type are the
+    // same. Currently only consider to eliminate rounding pairs for (bf16 / f16
+    // <-> f32).
+    auto input = truncFOp.getOperand();
+    auto inTy = input.getType();
+    auto outTy = extFOp.getType();
+    auto shortTy = getElementTypeOrSelf(truncFOp.getType());
+    if (inTy == outTy && getElementTypeOrSelf(inTy).isF32() &&
+        (shortTy.isF16() || shortTy.isBF16())) {
+      rewriter.replaceOp(extFOp, {input});
+      return success();
+    }
+
+    return failure();
+  }
+
+private:
+  FilterFunction filterFunc;
+};
+
+struct EliminateExplicitRounding final
+    : arith::impl::EliminateExplicitRoundingBase<EliminateExplicitRounding> {
+  using EliminateExplicitRoundingBase::EliminateExplicitRoundingBase;
+  void runOnOperation() override {
+    RewritePatternSet patterns(&getContext());
+    patterns.insert<EliminateExplicitRoundingRewritePattern>(&getContext());
+    FrozenRewritePatternSet patternSet(std::move(patterns));
+    SmallVector<Operation *> ops;
+    getOperation()->walk([&](arith::ExtFOp op) { ops.push_back(op); });
+    if (failed(applyOpPatternsAndFold(ops, patternSet)))
+      signalPassFailure();
```

Comment:
What kind of case need the greedy pattern rewriter instead of a simple IR walk and rewrite?

Reply:
If there are cases with nested trunc / ext op pairs, then a greedy pattern rewriter would be needed. However, in the current design, this pass is executed only after all f32 promotions are completed, so nested cases do not occur. The code snippet above is an example of a simple IR walk and rewrite.

Reply:
> However, in the current design, this pass is executed only after all f32 promotions are completed, so nested cases do not occur

This pass is not a part of a single pipeline; I don think we can assume this is the only context when it can be invoked. In fact, it's not part of any pass pipeline right now.

Reply:
> If there are cases with nested trunc / ext op pairs

Do you have an example? I don't picture an example which wouldn't work with a simple walk.

Reply:
Emm, I don't have an example, I think trunc / extf have no nested cases, so a simple walk should be enough.

Reply:
OK, so please remove the use of the greedy rewriter.

Reply:
I don't really get you here, I think the code snippet is a simple IR walk and rewrite. Could you please give me more insights or examples?

Reply:
A simple walk/rewrite does not involve `applyOpPatternsAndFold` nor a `RewritePatternSet`, instead it will look like:

```
getOperation()->walk([&](arith::ExtFOp op) {
    auto truncFOp = extFOp.getOperand().getDefiningOp<arith::TruncFOp>();
    if (!truncFOp)
      return;  
    ... // rewrite
}
```

Or if you'd like to use a pattern for some reason (why?):

```
EliminateExplicitRoundingRewritePattern pattern(&getContext());
PatternRewriter rewriter(&getContext());
getOperation()->walk([&](arith::ExtFOp op) {
  pattern.matchAndRewrite(op, rewriter);
}
```

Reply:
Thanks for your provided example, use the first code snippet instead.